account-compression/sdk/src/accounts/index.ts

```diff
@@ -0,0 +1,190 @@
+ */
+export type SplConcurrentMerkleTree = {
+  header: ConcurrentMerkleTreeHeader;
+  tree: ConcurrentMerkleTree;
+};
+
+type ConcurrentMerkleTreeHeader = {
+  accountType: number,
+  _padding: number[]
+  maxDepth: number; // u32
+  maxBufferSize: number; // u32
+  authority: PublicKey;
+  creationSlot: BN;
+};
+
+type ConcurrentMerkleTree = {
+  sequenceNumber: BN; // u64
+  activeIndex: number; // u64
+  bufferSize: number; // u64
+  changeLogs: ChangeLog[];
+  rightMostPath: Path;
+};
+
+export type PathNode = {
+  node: PublicKey;
+  index: number;
+};
+
+type ChangeLog = {
+  root: PublicKey;
+  pathNodes: PublicKey[];
+  index: number; // u32
+  _padding: number; // u32
+};
+
+type Path = {
+  leaf: PublicKey;
+  proof: PublicKey[];
+  index: number;
+  _padding: number;
+};
+
+export function deserializeConcurrentMerkleTree(buffer: Buffer): SplConcurrentMerkleTree {
+  let reader = new borsh.BinaryReader(buffer);
+
+  let header: ConcurrentMerkleTreeHeader = {
+    accountType: reader.readU8(),
+    _padding: Array.from(reader.readFixedArray(7)),
+    maxBufferSize: reader.readU32(),
+    maxDepth: reader.readU32(),
+    authority: readPublicKey(reader),
+    creationSlot: reader.readU64(),
+  };
+
+  let sequenceNumber = reader.readU64();
+  let activeIndex = reader.readU64().toNumber();
+  let bufferSize = reader.readU64().toNumber();
+
+  let changeLogs: ChangeLog[] = [];
+  for (let i = 0; i < header.maxBufferSize; i++) {
+    let root = readPublicKey(reader);
+
+    let pathNodes: PublicKey[] = [];
+    for (let j = 0; j < header.maxDepth; j++) {
+      pathNodes.push(readPublicKey(reader));
+    }
+    changeLogs.push({
+      pathNodes,
+      root,
+      index: reader.readU32(),
+      _padding: reader.readU32(),
+    });
+  }
+
+  // Decode Right-Most Path
+  let leaf = readPublicKey(reader);
+  let proof: PublicKey[] = [];
+  for (let j = 0; j < header.maxDepth; j++) {
+    proof.push(readPublicKey(reader));
+  }
+  const rightMostPath = {
+    proof,
+    leaf,
+    index: reader.readU32(),
+    _padding: reader.readU32(),
+  };
+
+  const tree = {
+    sequenceNumber,
+    activeIndex,
+    bufferSize,
+    changeLogs,
+    rightMostPath,
+  };
+
+  if (
+    getConcurrentMerkleTreeSize(header.maxDepth, header.maxBufferSize) !=
+    reader.offset
+  ) {
+
+    throw new Error(
+      "Failed to process whole buffer when deserializing Merkle Account Data"
+    );
+  }
+  return { header, tree };
+}
+
+export function getConcurrentMerkleTreeSize(
+  maxDepth: number,
+  maxBufferSize: number,
+  canopyDepth?: number
+): number {
+  let headerSize = 8 + 8 + 8 + 32;
+  let changeLogSize = (maxDepth * 32 + 32 + 4 + 4) * maxBufferSize;
+  let rightMostPathSize = maxDepth * 32 + 32 + 4 + 4;
+  let merkleRollSize = 8 + 8 + 8 + changeLogSize + rightMostPathSize;
+  let canopySize = 0;
+  if (canopyDepth) {
+    canopySize = ((1 << canopyDepth + 1) - 2) * 32
+  }
+  return headerSize + merkleRollSize + canopySize;
```

Comment:
how resistant is this to other changes? i wish there were a better way to express this byte layout...

Reply:
I rewrote this using the mocked-out layout of each individual component.

It will have to be manually synced with changes to the on-chain header struct and any changes made to the CMT library.